compiler-builtins/src/int/udiv.rs

```diff
@@ -43,6 +43,144 @@ intrinsics! {
 
         ((rem as u64) << 32) | (div as u64)
     }
+
+    /// Returns unsigned 8-bit `n / d` and `n % d`.
+    ///
+    /// Note: GCC implements a [non-standard calling convention](https://gcc.gnu.org/wiki/avr-gcc#Exceptions_to_the_Calling_Convention) for this function.
+    #[naked]
+    pub unsafe extern "C" fn __udivmodqi4() {
+        // Derived from: https://github.com/gcc-mirror/gcc/blob/95f10974a9190e345776604480a2df0191104308/libgcc/config/avr/lib1funcs.S#L1365
```

Comment:
These implementations derive from libgcc where the originating file is marked GPLv3+.  IANAL and unsure of the implications from including it here.

It would be interesting to attempt to port these functions to Rust.  I think a similar implementation may be possible using some of the standard library's u8/u16 methods.  I think that would require us to teach `rustc` about this weird ABI.  Does this implementation need to remain ABI compatible with GCC?

Reply:
Maybe instead of pulling GCC's impl we could simply use the naked function as a trampoline? i.e. like:

```rust
#[naked]
pub unsafe extern "C" fn __divmodhi4() {
    let out = /* call native-Rust implementation */;

    naked_asm!(/* move `out` to appropriate registers */)
}
```

This would be less efficient and we have to be careful not to overwrite registers we shouldn't touch, but it should be good enough to get started.

Reply:
I replied to this in another comment but unfortunately we can't take these as-is, ports are considered derivative work.

I mentioned here https://github.com/rust-lang/compiler-builtins/issues/711#issuecomment-2794732546 but adding a new ABI to rustc would be overkill for four functions in the whole ecosystem, maybe not possible based on LLVM IR limitations. You could probably use Rust as a start though, emit the optimized asm and manually adjust for the ABI. Or emit the optimized LLVM IR and adjust the signature to match the ABI if possible (this is effectively what compiler support in rustc would do), and compile that.